### DISCUSSION TITLE
How to Create your Skeleton Procedurally

### DISCUSSION THREAD

**Discussion by Fransferdy** [AndreaCatania/RuntimeSkeletalMeshGenerator#9] [General]
If you are trying to build procedural models with this, you will have to build your own skeleton (and most likely have your own code to animate the model).

By going deep into the engine FBX Importer code, I found how we can create an skeleton in runtime, it is fairly simple.
First, for some reason, the skeleton needs to have at least 2 bones.
Second the function that builds the skeleton cannot do anything else than create the skeleton, the reason for that, is that the skeleton is only actually built in the DESTRUCTOR of the Skeleton Creation utility class, so you'll setup the Skeleton Creation Utility class and only by the end of the function the skeleton will actually be created.

Your skeleton building function should have at least this signature(and maybe more parameters with your own bone/skeleton data objects):
```
void buildSkeleton(FReferenceSkeleton& refSkel, USkeleton* skeleton)
```
First you create an empty Skeleton object and call your function:
```
USkeleton* skeleton = NewObject<USkeleton>();

buildSkeleton((FReferenceSkeleton&)skeleton->GetReferenceSkeleton(), skeleton);
```

Include #include "BoneIndices.h" for the SkeletonCreation utility FReferenceSkeletonModifier.
The inside of buildSkeleton:
```
void buildSkeleton(FReferenceSkeleton& refSkel, USkeleton* skeleton)
{
	refSkel.Empty();
	FReferenceSkeletonModifier RefSkelModifier(refSkel, skeleton);

	const FMeshBoneInfo BoneInfo("bn0_in", "bn0_ex", INDEX_NONE);
	const FTransform BoneTransform(FTransform(FVector(0, 0, 0)));

	const FMeshBoneInfo BoneInfo2("bone1_in", "bone1_ex", 0);
	const FTransform BoneTransform2(FTransform(FVector(0, 50, 0)));

	RefSkelModifier.Add(BoneInfo, BoneTransform);
	RefSkelModifier.Add(BoneInfo2, BoneTransform2);
}
```

FReferenceSkeletonModifier is the class that will do the actual creation, from what I researched it is the only way to modify the UE internal skeleton/bone data.
RefSkelModifier.Add takes a BoneInfo struct and a FVector to create your bone,
BoneInfo parameters are "internal name"(of the bone), "external name"(the name it was in the file it was imported from, in our case, just the name with _ex appended) and the **parent** boneIndex. 
```
const FMeshBoneInfo BoneInfo("bn0_in", "bn0_ex", INDEX_NONE);
```
The root bone must have INDEX_NONE here, and subsequent bones have parent bone index 0...n.
The index of each bone will be defined by the order you add the bones with RefSkelModifier.Add .

So this is it for bone creation.

Regarding animation, you can send the output of GenerateSkeletalMeshComponent (a USkeletalMeshComponent*) to a PoseableMeshComponent in Blueprints. PoseableMeshComponent exposes functions that allow you to translate/rotate/transform each bone you defined, with this you can write your own code for animating your procedurally created Mesh!

**Comment by nathan-meshcapade** [comment]
Is it true that you can only do this in editor, or can you do this at runtime as well?

**Reply by Fransferdy** [reply]
This mini tutorial is how you do it in runtime, this one covers only how to make the skeleton, but then you have to assign each vertex a weight that corresponds to a a bone of that skeleton. After you have both skeleton and vertex "linked", you can proceed and move/transform those bones to create animations.

**Reply by lolaswift** [reply]
Have you looked into Unreal Engine 5.3 the new tools about mesh and skeleton? Any new ways of doing it? Is this still valid for ue5 as well? thanks.

**Reply by Fransferdy** [reply]
I am not sure what new tools for this use case are available in 5.3, but as others have noted in other pages around here, this solution (Runtime SkeletalMesh) works up to UE 5.1 and does not work for 5.2 and onwards. Additionaly, the original code here is full of bugs, you can check my fork, I added all community fixes there(but I'm not sure if it is free of bugs either).
